url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/genericServer/LocalStringConnection.java

```diff
@@ -0,0 +1,680 @@
+     *
+     * This class has a run method, but you must start the thread yourself.
+     * Constructors will not create or start a thread.
+     *
+     * @see #connect(String)
+     */
+    public LocalStringConnection()
+    {
+        in = new Vector();
+        out = new Vector();
+        init();
+    }
+
+    /**
+     * Constructor for an existing peer; we'll share two Vectors for in/out queues.
+     *
+     * This class has a run method, but you must start the thread yourself.
+     * Constructors will not create or start a thread.
+     *
+     * @param peer The peer to use.
+     *
+     * @throws EOFException If peer is at EOF already 
+     * @throws IllegalArgumentException if peer is null, or already
+     *   has a peer.
+     */
+    public LocalStringConnection(LocalStringConnection peer) throws EOFException
+    {
+        if (peer == null)
+            throw new IllegalArgumentException("peer null");
+        if (peer.ourPeer != null)
+            throw new IllegalArgumentException("peer already has a peer");
+        if (peer.isOutEOF() || peer.isInEOF())
+            throw new EOFException("peer EOF at constructor");
+
+        in = peer.out;
+        out = peer.in;
+        peer.ourPeer = this;
+        this.ourPeer = peer;
+        init();
+    }
+
+    /**
+     * Constructor common-fields initialization
+     */
+    private void init()
+    {
+        in_reachedEOF = false;
+        out_setEOF = false;
+        accepted = false;
+        data = null;
+        ourServer = null;
+        error = null;
+        connectTime = new Date();
+        appData = null;
+        remoteVersion = 0;
+        remoteVersionKnown = false;
+        remoteVersionTrack = false;
+    }
+
+    /**
+     * Read the next string sent from the remote end,
+     * blocking if necessary to wait.
+     *
+     * Synchronized on in-buffer.
+     * 
+     * @return Next string in the in-buffer
+     * @throws EOFException Our input buffer has reached EOF
+     * @throws IllegalStateException Server has not yet accepted our connection
+     */
+    public String readNext() throws EOFException, IllegalStateException
+    {
+        if (! accepted)
+        {
+            error = new IllegalStateException("Not accepted by server yet");
+            throw (IllegalStateException) error;
+        }
+        if (in_reachedEOF)
+        {
+            error = new EOFException();
+            throw (EOFException) error;
+        }
+
+        Object obj;
+
+        synchronized (in)
+        {
+            while (in.isEmpty())
+            {
+                if (in_reachedEOF)
+                {
+                    error = new EOFException();
+                    throw (EOFException) error;
+                }
+                else
+                {
+                    try
+                    {
+                        in.wait();
+                    }
+                    catch (InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)